File: src/main/java/my/study/repository/design/pattern/proto/java/WordDocument.java
package my.study.repository.design.pattern.proto.java;

import java.util.ArrayList;

/**
 * @author djh on  2018/11/28 16:40
 * @E-Mail 1544579459@qq.com
 */
public class WordDocument implements Cloneable {

    private String text;
    private ArrayList<String> imageList = new ArrayList<>();

    public WordDocument() {
        System.out.println("--------WordDocument的构造方法--------");
    }


    @Override
    protected WordDocument clone() throws CloneNotSupportedException {
        WordDocument cloneDocument = (WordDocument) super.clone();
        if (cloneDocument != null) {
            cloneDocument.text = text;
            cloneDocument.imageList = (ArrayList<String>) this.imageList.clone();
            return cloneDocument;
        }
        return null;
    }

    public void showDocument() {
        System.out.println("--------DocumentStart--------");
        System.out.println("Text: " + text);
        for (String s : imageList) {
            System.out.println("Image: " + s);
        }
        System.out.println("--------DocumentEnd--------");
    }

    public void setText(String text) {
        this.text = text;
    }

    public void addImage(String image) {
        imageList.add(image);
    }
}

File: src/main/java/my/study/repository/design/pattern/proto/kotlin/WordDocument.kt
package my.study.repository.design.pattern.proto.kotlin

/**
 * @author djh on  2018/11/28 17:11
 * @E-Mail 1544579459@qq.com
 */
class WordDocument : Cloneable {
    init {
        println("--------构造方法--------")
    }
    
    lateinit var text: String
    private var imageList = arrayListOf<String>()
    
    override fun clone(): WordDocument {
        val cloneDocument = super.clone() as WordDocument
        cloneDocument.text = text
        cloneDocument.imageList = imageList.clone() as ArrayList<String>
        return cloneDocument
    }
    
    fun addImage(image: String) {
        imageList.add(image)
    }
    
    fun showDocument() {
        println("--------DocumentStart--------")
        println("Text: $text")
        for (s in imageList) {
            println("Image: $s")
        }
        println("--------DocumentEnd--------")
    }
}